onnx/defs/controlflow/defs.cc

```diff
@@ -337,9 +337,9 @@ void LoopInferenceFunction(InferenceContext& ctx) {
 
       const bool is_loop_state_var = i < num_loop_state_vars;
 
-      if (!subgraph_output_type->has_tensor_type()) {
+      if (!subgraph_output_type->has_tensor_type() && !subgraph_output_type->has_sequence_type()) {
```

Comment:
Is there any reason (specific to the Loop op) to restrict the outputs to be of any type? E.g., why not maps? Or, sequences of sequences? (This is probably allowed by above check, but not by the type-constraint on the op below, which restricts it to sequences-of-tensors.) 

I think allowing the output-type to be unconstrained would require some extensions in the Schema implementation, which, I think, assumes that the set of possible types can be finitely enumerated and stored. It would be worth taking a look to see if this can be generalized. This would avoid having to generalize the Loop op again the next time if we want to allow other types.

Reply:
Today there does not exist operators that use/produce such flexible types. I agree to support types like sequence of sequence, we need to update Schema to relax the assumption of types being a finite set. However, I'm not sure if we should start this change with Loop, instead of a real use case. Should such op be added in the future, the infrastructure will have to be updated, and I'm relatively ok to generalize Loop (and If) accordingly in that opset update.

Reply:
The problem is that when that happens, we will have to bump the Loop operator's version again even though a Loop implementation itself does not really depend on the types of these values. This leads to unnecessary version-updates, which impacts exporters, convertors, backends. A loop is polymorphic over this type V and the constraints on the type V are artificial. (If this had been done correctly for the earlier version, even this extension would not have been needed).